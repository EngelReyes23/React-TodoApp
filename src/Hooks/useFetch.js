import { useEffect, useRef, useState } from "react";

export const useFetch = (url) => {
	// Referencias con useRef
	const isMount = useRef(true);

	// Effect
	useEffect(() => {
		return () => {
			isMount.current = false;
		};
	}, []);

	// state
	const [state, setState] = useState({
		data: null,
		loading: true,
		error: false,
	});

	useEffect(() => {
		if (isMount.current) {
			console.log("Fetch");
			setState({ data: null, loading: true, error: false });
			fetch(url)
				.then((res) => res.json())
				.then((data) => setState({ data, loading: false, error: false }))
				.catch(() => setState({ data: "Error", loading: true, error: true }));
		}
	}, [url]);

	return state;
};
