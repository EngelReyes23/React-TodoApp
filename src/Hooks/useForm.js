import { useState } from "react";

export const useForm = (initialValues = {}) => {
	//
	const [values, setValues] = useState(initialValues);

	// maneja los cambios en los inputs
	const handleInputChange = ({ target: { name, value } }) => {
		setValues({
			...values,
			[name]: value,
		});
	};

	return [values, handleInputChange];
};
