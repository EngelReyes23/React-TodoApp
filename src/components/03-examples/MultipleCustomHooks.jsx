import React from "react";
import useCounter from "../../Hooks/useCounter";
import { useFetch } from "../../Hooks/useFetch";
import "./styles.css";

const MultipleCustomHooks = () => {
	const { increment, decrement, counter, setCounter } = useCounter(1);
	const { loading, data } = useFetch(
		`https://breakingbadapi.com/api/quotes/${counter}`
	);

	const { quote, author } = !!data && data[0];

	return (
		<>
			<h1>BreakingBad Quotes</h1>
			<hr />
			{loading && (
				<div className="alert alert-info text-center">Loading...</div>
			)}

			{!loading && (
				<blockquote className="blockquote text-end">
					<p className="">{quote}</p>
					<footer className="blockquote-footer">{author}</footer>
				</blockquote>
			)}

			{counter !== 1 && (
				<button className="btn btn-primary m-1 shadow " onClick={decrement}>
					Previus
				</button>
			)}
			{counter !== 30 && (
				<button className="btn btn-primary m-1 shadow" onClick={increment}>
					Next
				</button>
			)}
			{counter === 30 && setCounter(63)}
		</>
	);
};

export default MultipleCustomHooks;
