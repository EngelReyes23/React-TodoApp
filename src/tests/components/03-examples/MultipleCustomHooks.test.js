import { shallow } from "enzyme";
import MultipleCustomHooks from "../../../components/03-examples/MultipleCustomHooks";
import { useFetch } from "../../../Hooks/useFetch";

// mock useFetch hook
jest.mock("../../../Hooks/useFetch");

describe("Pruebas en <MultipleCustomHooks />", () => {
	test("should mostrarse correctamente", () => {
		// simula el llamado a la api
		useFetch.mockReturnValue({ data: null, loading: true, error: false });

		const wrapper = shallow(<MultipleCustomHooks />);

		expect(wrapper).toMatchSnapshot();
	});

	test("should mostrarse correctamente con el valor de los inputs", () => {
		useFetch.mockReturnValue({
			data: [
				{
					author: "Pepito",
					quote: "Hola",
				},
			],
			loading: false,
			error: false,
		});

		const wrapper = shallow(<MultipleCustomHooks />);
		expect(wrapper.find(".alert").exists()).toBe(false);
		expect(wrapper.find("p").text().trim()).toBe("Hola");

		console.log(wrapper.html());
		expect(wrapper).toMatchSnapshot();
	});
});
