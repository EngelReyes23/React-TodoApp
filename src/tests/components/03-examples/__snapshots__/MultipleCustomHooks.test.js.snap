// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <MultipleCustomHooks /> should mostrarse correctamente 1`] = `
<Fragment>
  <h1>
    BreakingBad Quotes
  </h1>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    Loading...
  </div>
  <button
    className="btn btn-primary m-1 shadow"
    onClick={[Function]}
  >
    Next
  </button>
</Fragment>
`;

exports[`Pruebas en <MultipleCustomHooks /> should mostrarse correctamente con el valor de los inputs 1`] = `
<Fragment>
  <h1>
    BreakingBad Quotes
  </h1>
  <hr />
  <blockquote
    className="blockquote text-end"
  >
    <p
      className=""
    >
      Hola
    </p>
    <footer
      className="blockquote-footer"
    >
      Pepito
    </footer>
  </blockquote>
  <button
    className="btn btn-primary m-1 shadow"
    onClick={[Function]}
  >
    Next
  </button>
</Fragment>
`;
